test/unittests/t_util.cc

```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+  EXPECT_EQ("my;beautiful;string", JoinStrings(result, ";"));
+
+  result.push_back("");
+  EXPECT_EQ("mybeautifulstring", JoinStrings(result, ""));
+  EXPECT_EQ("my beautiful string ", JoinStrings(result, " "));
+  EXPECT_EQ("my;beautiful;string;", JoinStrings(result, ";"));
+}
+
+TEST_F(T_Util, ParseKeyvalMem) {
+  map<char, string> map;
+  string cvmfs_published =
+      "Cf3bc68897a32278da5b5b0e4b5e4711a9102dde5\n"
+      "B75834368\n"
+      "Zfirst\n"
+      "Zsecond\n"
+      "Rd41d8cd98f00b204e9800998ecf8427e\n"
+      "D900\n"
+      "S8418\n"
+      "Natlas.cern.ch\n"
+      "Ncms.cern.ch\n"
+      "X0b457ac12225018e0a15330364c20529e15012ab\n"
+      "H70a5de156ee5eaf4f8e191591b6ade378f1120bd\n"
+      "T1431669806\n"
+      "--\n"
+      "EEXTRA-CONTENT\n";
+  const unsigned char *buffer =
+      reinterpret_cast<const unsigned char *>(cvmfs_published.c_str());
+  ParseKeyvalMem(buffer, cvmfs_published.length(), &map);
+
+  EXPECT_EQ(10u, map.size());  // including the end() node in the map
+  EXPECT_EQ("f3bc68897a32278da5b5b0e4b5e4711a9102dde5", map['C']);
+  EXPECT_EQ("75834368", map['B']);
+  EXPECT_EQ("first|second", map['Z']);  // "first" shouldn't be overwritten
+  EXPECT_EQ("d41d8cd98f00b204e9800998ecf8427e", map['R']);
+  EXPECT_EQ("900", map['D']);
+  EXPECT_EQ("8418", map['S']);
+  EXPECT_EQ("cms.cern.ch", map['N']);  // "atlas.cern.ch" should be overwritten
+  EXPECT_EQ("0b457ac12225018e0a15330364c20529e15012ab", map['X']);
+  EXPECT_EQ("70a5de156ee5eaf4f8e191591b6ade378f1120bd", map['H']);
+  EXPECT_EQ("1431669806", map['T']);
+  EXPECT_TRUE(map.find('E') == map.end());
+}
+
+TEST_F(T_Util, ParseKeyvalPath) {
+  map<char, string> map;
+  const char *big_buffer = static_cast<const char *>(calloc(8000,
+      sizeof(char)));
+  string big_file = "bigfile.txt";
+  string content_file = "contentfile.txt";
+  string cvmfs_published =
+      "Cf3bc68897a32278da5b5b0e4b5e4711a9102dde5\n"
+      "B75834368\n"
+      "Zfirst\n"
+      "Zsecond\n"
+      "Rd41d8cd98f00b204e9800998ecf8427e\n"
+      "D900\n"
+      "S8418\n"
+      "Natlas.cern.ch\n"
+      "Ncms.cern.ch\n"
+      "X0b457ac12225018e0a15330364c20529e15012ab\n"
+      "H70a5de156ee5eaf4f8e191591b6ade378f1120bd\n"
+      "T1431669806\n"
+      "--\n"
+      "EEXTRA-CONTENT\n";
+  CreateFileWithContent(big_file, string(big_buffer));
+  CreateFileWithContent(content_file, cvmfs_published);
+
+  EXPECT_FALSE(ParseKeyvalPath("/path/that/does/not/exists.txt", &map));
+  EXPECT_FALSE(ParseKeyvalPath(string(sandbox) + "/" + big_file, &map));
+  EXPECT_TRUE(ParseKeyvalPath(string(sandbox) + "/" + content_file, &map));
+}
+
+TEST_F(T_Util, DiffTimeSeconds) {
+  struct timeval start = CreateTimeval(2, 900000);
+  struct timeval end = CreateTimeval(3, 200000);
+  EXPECT_EQ(0.3, DiffTimeSeconds(start, end));
+
+  start = CreateTimeval(8, 400000);
+  end = CreateTimeval(1, 800000);
+  EXPECT_NE(-6.6, DiffTimeSeconds(start, end));
```

Comment:
That the difference is not negative. In principle one would expect to find a negative difference, but the functions is not supposed to return a negative value. However it could be changed for EXPECT_GT(0, DiffTimeSeconds(start, end)